Implement --prune-whitelist(-w) flag to overwrite default whitelist for --prune

MrHohn opened the pull request:
From kubernetes/kubectl#1854.

Updates:

As suggested, the new commits implement a default whitelist for `kubectl apply --prune`, which could be overwritten by using `--prune-whitelist`or `-w` flag. The default whitelist and example as below.

Also supports `--dry-run` for `--prune` to fix kubernetes/kubernetes#35222.

whitelist:

```
type pruneResource struct {
    group      string
    version    string
    kind       string
    namespaced bool
}

*pruneResources = []pruneResource{
    {"", "v1", "ConfigMap", true},
    {"", "v1", "Endpoints", true},
    {"", "v1", "Namespace", false},
    {"", "v1", "PersistentVolumeClaim", true},
    {"", "v1", "PersistentVolume", false},
    {"", "v1", "Pod", true},
    {"", "v1", "ReplicationController", true},
    {"", "v1", "Secret", true},
    {"", "v1", "Service", true},
    {"batch", "v1", "Job", true},
    {"extensions", "v1beta1", "DaemonSet", true},
    {"extensions", "v1beta1", "Deployment", true},
    {"extensions", "v1beta1", "HorizontalPodAutoscaler", true},
    {"extensions", "v1beta1", "Ingress", true},
    {"extensions", "v1beta1", "ReplicaSet", true},
    {"apps", "v1beta1", "StatefulSet", true},
}
```

example:

```
$ kubectl apply -f /path/to/file --prune -l test=true \
      -w core/v1/Pod \
      -w core/v1/Service \
      -w extensions/v1beta1/Deployment 
```

@mikedanese

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35220)

<!-- Reviewable:end -->

mikedanese commented:
Thanks for picking this up. Would it make sense to just do all types always?

MrHohn commented:
Both ways make sense to me. But people might have other concerns I am not aware of.
Open to advice.

pwittrock commented:
I actually think we may want to solve this another way so it is more reproducible

mikedanese commented:
> I actually think we may want to solve this another way so it is more reproducible

Can you elaborate or sync with @MrHohn on how to achieve this?

pwittrock commented:
I need to come up with something more detailed.  The TLDR is that I think we want the options for kubectl apply to be stored in the directory that is applied.  `--prune`,  label selectors for matching resources, types of resources to prune, etc.  That way when you `apply` a directory the behavior isn't driven by the flags the user types in, but instead by the checked in configuration.

Config would
-  Specify an annotation applied to every resource it creates.  Only these resources would be deleted.  This would be used instead of the last-applied-config annotation.
- Prevent users from accidentally applying the wrong directory, or miss matching the directory and flags.
- Provide a way to manage how configuration is applied through checked in source instead of un reproducible commands.

May it is fine to skip this for now, but in that case --prune should only be used with CICD that ensures `apply` that is run is reproducible.

WDYT?

mikedanese commented:
I think we need a solution for v1.5 that isn't a massive time sink for me or @MrHohn who have other commitments for v1.5. I think it makes sense to apply --prune for all types and not add a flag until the overhaul you are describing. This is an expert feature that provides real value to users who understand the caveats and we should document it as so until we can add some safety and reproducibility guarantees. Can we label apply --prune as a nonstable API (essentially alpha) for v1.5?

pwittrock commented:
TLDR; That is fine lets document in the help that it is alpha and users must be prepared for backwards compatibility to break in subsequent releases.

I am not sure if we have a great way of communicating alpha features but yes we should definitely make this alpha.  Maybe in the help docs?  In 1.6 I would like --prune to fail if it cannot find a config.  Maybe in 1.6 prune should just require a path to the prune config or something.  Alternatively we could make it a subcommand instead so prune-specific flags.

MrHohn commented:
Thanks for the discussions. I will soon change it to prune all types as a temporary solution.

pwittrock commented:
WDYT of having a default whitelist that can be overridden with a flag?

MrHohn commented:
Sounds reasonable. What about putting `core` and `extensions` groups into whitelist?

MrHohn commented:
Though the `prune` feature are becoming stranger after this modification :(

Sorry for messing this up, but the addon manager really need this feature.

MrHohn commented:
The new commits implement the default whitelist and the overridden flag. The detail is updated on the first post.

Currently the whitelist is hardcoded. I try to avoid hardcoding by using package like [the discovery client](https://github.com/kubernetes/kubernetes/tree/master/pkg/client/typed/discovery). But somehow the types return by server are also need to be manually filtered, some of them either do not have valid RESTMapping, or have RESTMapping but the server return error `could not find the resource`, or the server does not allow to list those resources. As it became more complex to distinguish those situations, I hardcoded the whitelist.

MrHohn commented:
This is blocking #33698.

mikedanese commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -123,6 +128,27 @@ func validatePruneAll(prune, all bool, selector string) error {
 	return nil
 }
 
+func validatePruneResources(pruneResources *[]pruneResource, gvks []string) error {
```
It's weird that this function is called `validatePruneResources` but it's actually modifying it's input argument. Is it doing to much or should it be renamed?

MrHohn replied:
Thanks for the comment. I changed the name to `parsePruneResources` and put it into RunApply().

mikedanese commented:
@fabianofranz @pwittrock PTAL. I'm pretty unfamiliar with kubectl so not sure if there is a better way to do this. This looks fine to me.

mikedanese commented:
@fabianofranz @pwittrock ping.

pwittrock commented:
Generally looks good.  Comments are generally minor except for the StatefulSets pruning.

@mikedanese I will be OOO for the next 2 weeks with limited availability.  Would you be willing to close off any of the few comments I left?  Otherwise LGTM.

---

Reviewed 1 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 131 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KUYIsvD6cPHOBLnOVkg-r2-131:-KVVuvZ0IPGZRtq-Atfx:bbk5srr) ([raw file](https://github.com/kubernetes/kubernetes/blob/4b8676a13a7b576f6d13b78d05220e11d5276197/pkg/kubectl/cmd/apply.go#L131)):*
<details><summary><i>Previously, MrHohn (Zihong Zheng) wrote…</i></summary>
> Thanks for the comment. I changed the name to `parsePruneResources` and put it into RunApply().
</details>
Non-blocker: Why not just return a slice with the error instead of passing it in?

---

*[pkg/kubectl/cmd/apply.go, line 334 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVveeXD02helgwFH44:-KVVveeXD02helgwFH45:bu9dfu3) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L334)):*
> ```Go
> 
> func (pr pruneResource) String() string {
> 	return fmt.Sprintf("%v/%v, Kind=%v, namespaced=%v", pr.group, pr.version, pr.kind, pr.namespaced)
> ```

nit: any particular reason Kind is capitalized while namespaced is not?

---

*[pkg/kubectl/cmd/apply.go, line 341 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvA93YYGaqjl9o7wL:-KVVvA93YYGaqjl9o7wM:bw7r4ya) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L341)):*
> ```Go
> 		// default whitelist
> 		*pruneResources = []pruneResource{
> 			{"", "v1", "ConfigMap", true},
> ```

What about StatefulSet?  Anything else new to beta or ga in 1.5?

---

*[pkg/kubectl/cmd/apply.go, line 350 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvJFAHmEg2VYQTz4V:-KVVvJFAHmEg2VYQTz4W:b-d8zksg) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L350)):*
> ```Go
> 			{"", "v1", "Secret", true},
> 			{"", "v1", "Service", true},
> 			{"batch", "v1", "Job", true},
> ```

What should we do with older api versions - e.g. v1beta1 jobs?  At least add a TODO about this and file an issue.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-:-KVVy-uPzrTDY5DNBdcQ:b-8f3ztk)*
<!-- Sent from Reviewable.io -->

mikedanese commented:
@pwittrock Thanks for taking a look and giving feedback. I will be happy to finish the review.

MrHohn commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 131 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KUYIsvD6cPHOBLnOVkg-r2-131:-KVWG-UcGdO7HuXhqzjS:b10suy0) ([raw file](https://github.com/kubernetes/kubernetes/blob/4b8676a13a7b576f6d13b78d05220e11d5276197/pkg/kubectl/cmd/apply.go#L131)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Non-blocker: Why not just return a slice with the error instead of passing it in?
</details>
Sure, changed.

---

*[pkg/kubectl/cmd/apply.go, line 334 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVveeXD02helgwFH44:-KVWGAlXNKw8ZpkkVWvD:bt2vr1b) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L334)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> nit: any particular reason Kind is capitalized while namespaced is not?
</details>
Fixed.

---

*[pkg/kubectl/cmd/apply.go, line 341 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvA93YYGaqjl9o7wL:-KVWGBshRBJJxA6G24G4:b-sxz49v) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L341)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What about StatefulSet?  Anything else new to beta or ga in 1.5?
</details>
Added StatefulSet.

---

*[pkg/kubectl/cmd/apply.go, line 350 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220#-KVVvJFAHmEg2VYQTz4V:-KVWGF5ZZWhArslmesT9:bb3rkjp) ([raw file](https://github.com/kubernetes/kubernetes/blob/858fa1a85c99c66234f67047d75e4bcf6dd35f0f/pkg/kubectl/cmd/apply.go#L350)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What should we do with older api versions - e.g. v1beta1 jobs?  At least add a TODO about this and file an issue.
</details>
Issue filed on #35991.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35220)*
<!-- Sent from Reviewable.io -->

MrHohn commented:
Many thanks for reviewing this! New commits are pushed.

Notice that `{"apps", "v1beta1", "StatefulSet", true}` is used for the StatefulSet, but PR #35731 that moves `StatuefulSet` from `v1alpha1` to `v1beta1` is umerged yet. However, it should not block this PR because for registeredMapper.RESTMapping() the version argument is just a preference but not mandatory.

In this case, RESTMapper will instead return `apps/v1alpha1, Kind=StatefulSet` if it does not have `v1beta1` yet. As below logs I got:
```
# use apps/v1beta1, Kind=StatefulSet
$ kubectl apply -f /a.yaml --prune -l prune-group=true --dry-run
pod "a" created (dry run)
...
[debug] apps/v1alpha1, Kind=StatefulSet
```

fabianofranz commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -271,26 +297,83 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *App
+	for _, resource := range *pruneResources {
+		addedMapping, err := registeredMapper.RESTMapping(unversioned.GroupKind{Group: resource.group, Kind: resource.kind}, resource.version)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid resrouce %v: %v", resource, err)
```
typo "resrouce"

fabianofranz commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -46,6 +48,7 @@ type ApplyOptions struct {
 	Prune           bool
 	Cascade         bool
 	GracePeriod     int
+	pruneResources  []pruneResource
```
Any reason why this is not exposed? Would make it better testable.

fabianofranz commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -101,6 +104,7 @@ func NewCmdApply(f cmdutil.Factory, out io.Writer) *cobra.Command {
 	cmdutil.AddValidateFlags(cmd)
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on")
 	cmd.Flags().Bool("all", false, "[-all] to select all the specified resources.")
+	cmd.Flags().StringArrayP("groupversionkind", "g", []string{}, "Overwrite the default whitelist with <group/version/kind> for --prune")
```
We kind of try to have dashes in flag names as a convention. Given you have a short form, could this be `--group-version-kind`, or something directly related to the purpose instead of the format, like `--prune-whitelist`, `--skip-when-pruning`, `--prune-skipped`, or something similar? Starting with `--prune*` would also make them appear together in the flags list.  ;)

MrHohn replied:
Make sense. Changing it to `--prune-whitelist`

fabianofranz commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -135,6 +161,11 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *App
 		return err
 	}
 
+	options.pruneResources, err = parsePruneResources(cmdutil.GetFlagStringArray(cmd, "groupversionkind"))
```
Is this flag mutually exclusive with another one, like `--all`? For example, what happens if you provide both `--all` and `--groupversionkind`? You may need to validate if they're mutually exclusive, or make it clear if one overrides the behavior of the other.

MrHohn replied:
I think `--all` and `--gourpversionkind` should be allowed to appear together? As the former indicates all objects regardless to label selector and the latter indicates the resource type?

fabianofranz reviewed (commented):
A few comments. Sorry about the delay!

MrHohn committed:
Implements --prune-whitelist(-w) flag to overwrite default whitelist for --prune

MrHohn committed:
Adds a test for apply --prune --prune-whitelist

MrHohn commented:
@k8s-bot unit test this, issue: #35872

MrHohn commented:
From GCE e2e test:
```
fatal: unable to access 'https://github.com/kubernetes/release/': Could not resolve host: github.com
```

Seems unrelated.
@k8s-bot cvm gce e2e test this, issue: #IGNORE

MrHohn commented:
The new commit changes `--groupversionkind`(`-g`) flag to `--prune-whitelist`(`-w`).

MrHohn commented:
Codes were modified according to the instructions. Is it ready to go? @fabianofranz @mikedanese 

This is also blocking #33239.

mikedanese commented:
LGTM

MrHohn commented:
Many thanks!

bprashanth commented:
@pwittrock I haven't been following this too closely, but regarding https://github.com/kubernetes/kubernetes/pull/35220#issuecomment-255236360

>TLDR; That is fine lets document in the help that it is alpha and users must be prepared for backwards compatibility to break in subsequent releases.

If you want it to truly be "alpha", I think we will have to flag control it, and disable it on platforms with a support contract (and think of another way to handle addon update by monday). 

I assume you mean this is effectively beta (i.e works on GKE).

bprashanth commented:
Actually I don't think that matters. 
I don't even know what the guarantee is for client side alpha features against GKE clusters. If someone starts using it, and it works, it'll keep working till they upgrade the client. They're 100% in control, they just can't use 1.6 kubectl features AND have this feature.
